### ISSUE TITLE
Crazy proposal to make it feasible to allow external stylesheets

### ISSUE THREAD

**Issue by pbakaus** [ampproject/amphtml#4609]
Hey folks,

I just had a pretty crazy idea that would allow us to allow external stylesheets without the big penalty associated with it on high latency situations. Here's how it works:
1. page author normally includes a stylesheet in the header
2. when the AMP cache fetches the page, it stores two versions of the page: the first one is straight forward a copy, and the second one is a copy that has the stylesheet inlined
3. when a user gets AMP search results in Search (for example), the AMP runtime, during prerender phase makes an additional HEAD http request to the linked stylesheet to find out if the stylesheet is browser-cached or not (if 203, it's not in the cache yet).
4. If the stylesheet is not found in the browser cache: AMP Cache serves the pre-inlined version of the page
5. If the stylesheet is found in the browser cache: AMP Cache serves the normal version with the external asset, as it knows there's no additional HTTP request necessary to resolve it.

In order to implement this, we would need the following:
1. We would need to know about the linked stylesheet when serving AMP pages _before_ the actual page is served, so it needs to returned as part of the Cache API
2. We need to store two copies of every page that has external links in the cache.

But! This could give us the best of both worlds – very low friction initial serving, less bandwidth servings of pages of the same original origin upon subsequent navigation.

Thoughts?

**Comment by dknecht** [comment]
In which situation would the stylesheet ever get cached?

I thought the issue with large stylesheets was latency and also CPU / power requirements for a mobile browser to compute the larger complex stylesheets.

**Comment by dvoytenko** [comment]
There are two issues:

1. Controlling the size of CSS
2. Avoiding sequential blocks to read CSS before the page can be styled.

CDN can obviously inline the external CSS. But there are also very many vulcanizing tools out there that do the same with ease.

Also, we are working with browsers to come up with a resource size spec that could allow us to do things like externalizing CSS.

**Comment by pbakaus** [comment]
@dknecht the stylesheet would be lazy loaded (out of DOM) the first time right after the inlined doc is fetched. Still low latency when the browser cache is hit, and we could keep the size restraint.

@dvoytenko 

1) we can solve this easily, we'll just have the validator fetch the CSS and validate size that way. Super simple (maybe a little more expensive of course).
2) Vulcanize doesn't help solve the problem that my proposal solves – by vulcanizing, you circumvent the browser cache and have to fetch the CSS over and over for the same domain, on every request. This means that a Service Worker cache will be super heavy, for instance (ampproject.org's SW is 15mb!).

So far, I'm not convinced I heard anything that is a blocker for my idea :) The biggest issue I've heard is that this will double storage requirements for the cache, as the cache has two keep two versions of every page.

**Comment by dvoytenko** [comment]
@pbakaus Ok, I think I misunderstood your proposal then.

> the AMP runtime, during prerender phase makes an additional HEAD http request to the linked stylesheet to find out if the stylesheet is browser-cached or not (if 203, it's not in the cache yet)

It will send a request to the origin server? Why do we expect 203 to be received?

> If the stylesheet is not found in the browser cache: AMP Cache serves the pre-inlined version of the page

Since you describe above "AMP runtime ... makes an additional HEAD request" I assume that AMP page is already loaded? I mean how would Cache know what to serve on the initial request?

**Comment by pbakaus** [comment]
No, the AMP page wouldn't be already loaded:

1. When the cache saves a copy of the AMP page, it also saves info on the URL of the linked stylesheet.
2. Later on, when a viewer sends a bunch of canonical URLs to the Cache API to find out if there are AMP versions associated (to then preload them), part of the JSON is the URL of the stylesheet, so the stylesheet can be probed in advance without loading the rest of the page.

There is, however, probably a better way to check whether something is cached or not that doesn't require an additional HTTP request (even though it's just a HEAD request) every time we warm up for prerender.

In [this article](https://css-tricks.com/case-study-boosting-front-end-performance/#article-header-id-19), the author implements a very similar idea using Cookies – since we're all client side, local storage might work just as well for us.

Basically, the flow when rendering an AMP page looks like this:

1. We check in the users' local storage (or cookies) to see if the stylesheet of the current page is cached or not. If it isn't:
2. We serve an AMP page that has the stylesheet inlined, while simultaneously lazy loading the stylesheet separately, just so it's in the browser cache.
3. We set a cookie or local storage entry that says "yes, that stylesheet is now in the browser cache").
4. Next time the user opens a link from the same origin, the cache serves the standard AMP page, with the external stylesheet, as it confidently knows it's in the browser cache already.

The final option to make this all happen is by utilizing a Service Worker on the cache.

**Comment by dvoytenko** [comment]
@pbakaus And is the stylesheet loaded via proxy or from the origin?

**Comment by pbakaus** [comment]
@dvoytenko via proxy, we'd give it some distinct version in the name, like so:

Original stylesheet url found on the page is:

```
https://ampproject.org/style.css
```

Proxy downloads it, generates a checksum, and produces the following carbon copy:

```
https://cdn.ampproject.org/c/style.css?v=48rf2f02
```

by using the distinct checksum, we can set the browser cache to never expire, and then reliably set cookie/local storage.

**Comment by cramforce** [comment]
This is very similar to solving the problem "Should I HTTP2 push this resource or does the client already have it cached?" which is a really hard problem in practice.

I don't actually think that this type of "in in cache" stuff via XHR is possible. The status code is always the same. The only way to implement this in a ServiceWorker.

I think a more feasible approach would be more an AMP Cache to decide to un-inline a stylesheet and initially always HTTP2 push it with the doc (the client can try to cancel the push).

**Comment by pbakaus** [comment]
In that case, let's just remove the inlining requirement altogether? HTTP2 push will be sufficient as alternative.

**Comment by adelinamart** [comment]
@pbakaus are there still things to discuss here? Thanks

**Comment by pbakaus** [comment]
There are things still do do here :) Would be nice if we could find a way to remove the inlining limitation. But we can close this one for now and revisit at a later point.

**Comment by varunwinni** [comment]
I want to access external style sheet please check once 

http://stackoverflow.com/questions/42199925/external-style-sheet-is-not-working-in-amp-pages
